### ISSUE TITLE
Annotation to indicate arguments are optional

### ISSUE THREAD

**Issue by jasom** [koalaman/shellcheck#703]
I have a couple of utility functions that take an optional $1, and read from stdin otherwise.  I'd _really_ like to not have to do a `#shellcheck disable=SC2119` every single time I use them.  If we could have something like `#shellcheck command_info=optional_args` that I could put right above the command definition, it would greatly reduce the noise from shellcheck.

**Comment by koalaman** [comment]
You can actually put the annotation before the function definition, and it will apply to the whole function! Does that solve the problem?

**Comment by FloGa** [comment]
Since this issue seems not to be solved, I'll try it with a minimal example:

```bash
#!/bin/bash
relax() {
  local duration="${1:-2}"
  echo "Sleeping for $duration seconds ..."
  sleep "$duration"
}

# do something
relax
```

This will trigger SC2119 as well as SC2120. But `$1` in `relax` is clearly optional. If unset, it will be assigned to `2`. It would be nice if one could tell shellcheck to ignore warnings regarding this optional parameter.

If I ignore SC2119 globally, I will miss *actual* mistakes I might have made.

**Comment by matthewleon** [comment]
@koalaman would you be open to an approach to solving the issue above that involves shellcheck detecting that the argument in question is optional? In some cases, like the example given above, where there is a default value, this should be fixable without resorting to annotation. Admittedly, more advanced cases would be trickier...

**Comment by Dabombber** [comment]
> 
> 
> You can actually put the annotation before the function definition, and it will apply to the whole function! Does that solve the problem?

This seems to be broken, although it works if there's no variables in the script.
```shell
#!/bin/sh

VAR="value"

# shellcheck disable=SC2120
# shellcheck disable=SC2119
foo() {
	[ -z "$*" ] && echo "hello"
}

bar() {
	echo "$VAR"
	foo
}
```

**Comment by koalaman** [comment]
@Dabombber I see what you mean. It fails because the SC2119 is emitted in `bar` and not in `foo`. I don't have a good suggestion for that, other than maybe squashing it filewide which is not great :|

**Comment by koalaman** [comment]
As of the latest revision, disabling SC2120 on a function will also disable SC2119 on each call site. Thanks!

I have not gone through to find other companion warnings that should be treated similarly, but there are undoubtedly some.
